File: src/parser.c
#include "parser.h"

#include "allocator.h"
#include "ast.h"

#include <stdarg.h>
#include <stdlib.h>
#include <stdnoreturn.h>

VECTOR_IMPL(struct ast_property*, ast_property_ptr)

static noreturn void
vsynerror(
    struct allocator allocator[static const 1], struct token token, va_list ap
) {
    enum lex_token_type type = va_arg(ap, enum lex_token_type);

    fprintf(
        stderr, STR_FMT ":%d:%d: syntax error: expected ",
        STR_ARG(sources[token.location.file]), token.location.line_number,
        token.location.column_number
    );

    while (type != TOKEN_EOF) {
        if (type == TOKEN_INTEGER || type == TOKEN_NAME) {
            fprintf(stderr, STR_FMT, STR_ARG(token_names[type]));
        } else {
            fprintf(stderr, "'" STR_FMT "'", STR_ARG(token_names[type]));
        }

        type = va_arg(ap, enum lex_token_type);
        fprintf(stderr, ", ");
    }

    fprintf(stderr, "found '" STR_FMT "'\n", STR_ARG(token_names[token.type]));
    error_line(allocator, token.location);
    exit(EXIT_PARSE);
}

static noreturn void
synerror(struct allocator allocator[static const 1], struct token token, ...) {
    va_list ap;
    va_start(ap, token);
    vsynerror(allocator, token, ap);
    va_end(ap);
}

static void
synassert(
    struct allocator allocator[static const 1], bool condition,
    struct token token, ...
) {
    if (!condition) {
        va_list ap;
        va_start(ap, token);
        vsynerror(allocator, token, ap);
        va_end(ap);
    }
}

static void
expect_token(
    struct lexer lexer[static const 1], enum lex_token_type token_type,
    struct token token[const 1]
) {
    struct token _token = {};
    struct token* out   = token != nullptr ? token : &_token;
    lex(lexer, out);
    synassert(
        lexer->allocator, out->type == token_type, *out, token_type, TOKEN_EOF
    );
    if (token == nullptr) {
        token_finish(out);
    }
}

static struct ast_property* parse_property(
    struct parser_context parser_context[static const 1],
    struct lexer lexer[static const 1]
);

static struct ast_element*
parse_element(
    struct parser_context parser_context[static const 1],
    struct lexer lexer[static const 1]
) {
    struct token token          = {};
    struct ast_element* element = allocator_allocate(
        parser_context->allocator, sizeof(struct ast_element)
    );

    switch (lex(lexer, &token)) {
        case TOKEN_NAME:
            element->type = AST_ELEMENT_TYPE_ID;
            element->id.id
                = string_duplicate(parser_context->allocator, token.name);
            break;
        case TOKEN_OPEN_BRACE:
            struct token token2 = {};
            bool break_out      = false;
            element->type       = AST_ELEMENT_TYPE_PROPERTIES;
            element->properties
                = vector_ast_property_ptr_new(parser_context->allocator);
            while (!break_out) {
                switch (lex(lexer, &token2)) {
                    case TOKEN_CLOSE_BRACE:
                        break_out = true;
                        break;
                    default:
                        unlex(lexer, &token2);
                        struct ast_property* next
                            = parse_property(parser_context, lexer);
                        vector_ast_property_ptr_add(&element->properties, next);
                        break;
                }
            }
            break;
        case TOKEN_INTEGER:
            element->type          = AST_ELEMENT_TYPE_INTEGER;
            element->integer.value = token.integer;
            break;
        case TOKEN_STRING:
            element->type = AST_ELEMENT_TYPE_STRING;
            element->string.value
                = string_duplicate(parser_context->allocator, token.string);
            break;
        default:
            unlex(lexer, &token);
            break;
    }

    return element;
}

static struct ast_property*
parse_property(
    struct parser_context parser_context[static const 1],
    struct lexer lexer[static const 1]
) {
    struct token token       = {};
    enum lex_token_type type = lex(lexer, &token);
    if (type != TOKEN_NAME) {
        unlex(lexer, &token);
        return nullptr;
    }
    struct ast_property* property = allocator_allocate(
        parser_context->allocator, sizeof(struct ast_property)
    );
    property->id = parser_context->next_id;
    parser_context->next_id += 1;
    property->name = string_duplicate(parser_context->allocator, token.name);
    expect_token(lexer, TOKEN_ASSIGN, nullptr);
    property->value = parse_element(parser_context, lexer);
    expect_token(lexer, TOKEN_SEMICOLON, nullptr);

    hashmap_properties_add(&parser_context->properties, property->id, property);

    return property;
}

static struct ast_element*
parse_root(
    struct parser_context parser_context[static const 1],
    struct lexer lexer[static const 1]
) {
    struct ast_element* root = allocator_allocate(
        parser_context->allocator, sizeof(struct ast_element)
    );
    root->type       = AST_ELEMENT_TYPE_PROPERTIES;
    root->properties = vector_ast_property_ptr_new(parser_context->allocator);

    for (;;) {
        struct ast_property* property = parse_property(parser_context, lexer);
        if (property == nullptr) {
            break;
        }
        vector_ast_property_ptr_add(&root->properties, property);
    }

    return root;
}

struct ast_element*
parse(
    struct parser_context parser_context[static const 1],
    struct lexer lexer[static const 1]
) {
    struct ast_element* root = parse_root(parser_context, lexer);
    expect_token(lexer, TOKEN_EOF, nullptr);
    return root;
}

File: src/parser.zig
const std = @import("std");

const ast = @import("ast.zig");
const AstProperty = ast.AstProperty;
const AstElement = ast.AstElement;
const lex = @import("lexer.zig");
const Lexer = lex.Lexer;
const LexTokenType = lex.LexTokenType;
const Token = lex.Token;
const util = @import("util.zig");
const Location = util.Location;
const Type = @import("types.zig").Type;

fn vsynerror(allocator: std.mem.Allocator, token: Token, args: std.ArrayList(LexTokenType)) anyerror!noreturn {
    const stderr = std.io.getStdErr().writer();

    try stderr.print("{s}:{}:{}: syntax error: expected ", .{
        util.sources.items[token.location.file],
        token.location.line_number,
        token.location.column_number,
    });

    var i: usize = 0;
    while (i < args.items.len) : (i += 1) {
        const token_type = args.items[i];
        try stderr.print("{s}, ", .{@tagName(token_type)});
    }

    try stderr.print("found '{s}'\n", .{@tagName(token.type)});
    try token.location.errorLine(allocator);
    std.process.exit(1);
}

fn synassert(allocator: std.mem.Allocator, token: Token, condition: bool, args: std.ArrayList(LexTokenType)) anyerror!void {
    if (!condition) {
        try vsynerror(allocator, token, args);
    }
}

pub const Parser = struct {
    allocator: std.mem.Allocator,
    next_id: usize,
    properties: std.AutoHashMap(usize, *AstProperty),
    lexer: *Lexer,

    pub fn init(allocator: std.mem.Allocator, lexer: *Lexer) @This() {
        return .{
            .allocator = allocator,
            .next_id = 1,
            .properties = std.AutoHashMap(usize, *AstProperty).init(allocator),
            .lexer = lexer,
        };
    }

    pub fn parse(self: *@This()) !*AstElement {
        const root = try self.parseRoot();
        try self.expectToken(.eof, null);
        return root;
    }

    fn parseRoot(self: *@This()) !*AstElement {
        var root = try self.allocator.create(AstElement);
        root.value = .{
            .properties = std.ArrayList(*AstProperty).init(self.allocator),
        };

        while (true) {
            const property = try self.parseProperty();
            if (property == null) {
                break;
            }
            try root.value.properties.append(property.?);
        }

        return root;
    }

    fn parseProperty(self: *@This()) !?*AstProperty {
        var token = Token.initEmpty();
        const token_type = try self.lexer.lex(&token);
        if (token_type != .name) {
            self.lexer.unlex(&token);
            return null;
        }

        var property = try self.allocator.create(AstProperty);
        property.id = self.next_id;
        self.next_id += 1;
        property.name = try self.allocator.dupe(u8, token.value.name);
        try self.expectToken(.assign, null);
        property.value = try self.parseElement();
        try self.expectToken(.semicolon, null);

        try self.properties.put(property.id, property);

        return property;
    }

    fn parseElement(self: *@This()) anyerror!*AstElement {
        var token = Token.initEmpty();
        var element = try self.allocator.create(AstElement);
        element.value = .{
            .integer = 0,
        };

        const token_type = try self.lexer.lex(&token);
        switch (token_type) {
            .name => {
                element.value = .{
                    .property_access = std.ArrayList([]const u8).init(self.allocator),
                };
                const name = try self.allocator.dupe(u8, token.value.name);
                try element.value.property_access.append(name);
                var token2 = Token.initEmpty();
                loop: while (true) {
                    const token_type2 = try self.lexer.lex(&token2);
                    switch (token_type2) {
                        .dot => {
                            try self.expectToken(.name, &token2);
                            const name2 = try self.allocator.dupe(u8, token2.value.name);
                            try element.value.property_access.append(name2);
                        },
                        else => {
                            self.lexer.unlex(&token2);
                            break :loop;
                        },
                    }
                }
            },
            .open_brace => {
                element.value = .{
                    .properties = std.ArrayList(*AstProperty).init(self.allocator),
                };
                loop: while (true) {
                    var token2 = Token.initEmpty();
                    const token_type2 = try self.lexer.lex(&token2);
                    switch (token_type2) {
                        .close_brace => {
                            break :loop;
                        },
                        else => {
                            self.lexer.unlex(&token2);
                            const next = try self.parseProperty();
                            try element.value.properties.append(next.?);
                        },
                    }
                }
            },
            .integer => {
                element.value = .{
                    .integer = token.value.integer,
                };
            },
            .string => {
                element.value = .{
                    .string = try self.allocator.dupe(u8, token.value.string),
                };
            },
            else => {
                self.lexer.unlex(&token);
            },
        }

        return element;
    }

    fn expectToken(self: *@This(), token_type: LexTokenType, token: ?*Token) anyerror!void {
        var emptyToken = Token.initEmpty();
        var out: *Token = if (token) |t| t else &emptyToken;
        _ = try self.lexer.lex(out);

        var tokens = std.ArrayList(LexTokenType).init(self.allocator);
        try tokens.append(token_type);

        try synassert(self.allocator, out.*, out.type == token_type, tokens);
        if (token == null) {
            out.tokenFinish();
        }
    }
};
